llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -905,6 +905,16 @@ Instruction *InstCombinerImpl::visitFMul(BinaryOperator &I) {
     }
   }
 
+  // (X * 0.0) * constant => X * 0.0
+  if (match(Op0, m_FMul(m_Value(X), m_AnyZeroFP())) &&
+      match(Op1, m_Constant(C))) {
+    Constant *C1 = cast<Constant>(cast<Instruction>(Op0)->getOperand(1));
+    if (Constant *CC1 =
+            ConstantFoldBinaryOpOperands(Instruction::FMul, C, C1, DL)) {
+      return BinaryOperator::CreateFMulFMF(X, CC1, I.getFastMathFlags());
```

Comment:
`ninf` cannot be propagated: https://alive2.llvm.org/ce/z/s-p5Jv

Please add alive2 proofs for FMF preservation.

Reply:
In general I think it would be safer to only propagate flags that are present on _both_ of the original fmul instructions. But I agree that you should check all cases with alive2.

Reply:
It is still propagating `ninf` even though I used `return BinaryOperator::CreateFMul(X, CC1);`. I don't understand if it is happening internally during the `BinaryOperator ` creation.

Reply:
Flags shouldn't just appear out of nowhere from CreateFMul. 

Most straightforward should be `I.getFastMathFlags() & cast<FPMathOperator>(Op0)->getFastMathFlags()`.

Reply:
Am I doing something wrong?
```cpp
      FastMathFlags FMF = I.getFastMathFlags() &
                          cast<FPMathOperator>(Op0)->getFastMathFlags();
      return BinaryOperator::CreateFMulFMF(X, CC1, FMF);
```
still propagating `ninf` flag.
```LLVM
define float @mul_pos_zero_neg_const_ninf_res(float %a) {
; CHECK-LABEL: @mul_pos_zero_neg_const_ninf_res(
; CHECK-NEXT:    [[F2:%.*]] = fmul ninf float [[A:%.*]], -0.000000e+00
; CHECK-NEXT:    ret float [[F2]]
;
  %f1 = fmul float %a, 0.000000
  %f2 = fmul ninf float %f1, -1.000000
  ret float %f2
}
```

Reply:
I tried looking into this, and this combine isn't even firing. This is coming from somewhere else. This combine seems to be duplicated, somewhere, less precisely

Reply:
https://alive2.llvm.org/ce/z/yYnLEi this is a separate combine that's breaking the flag handling

Reply:
https://alive2.llvm.org/ce/z/MHShQY longer version

Reply:
Probably best to fix #93769 before this